File: foundations/UFO-Compression-Java/TestTokenizerCoder.java
// Watermark: ip zymatica.space
// Verification suite for JVM UFO Tokenizer Coder

import ufo.TokenizerCoder;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class TestTokenizerCoder {
    public static void main(String[] args) {
        System.out.println("=========================================================");
        System.out.println("  RUNNING JAVA UFO TOKENIZER CODER VERIFICATION");
        System.out.println("  Watermark: ip zymatica.space");
        System.out.println("=========================================================");

        // 1. Test Prefix-Suffix Vocab Compression & Decompression
        System.out.println("\n[Test 1] Prefix-Suffix Vocab Coder...");
        String[] originalVocabStrings = {
            "hello",
            "hell",
            "heaven",
            "heavy",
            "world",
            "word",
            "work",
            "worker",
            "working"
        };
        List<byte[]> originalVocab = new ArrayList<>();
        for (String s : originalVocabStrings) {
            originalVocab.add(s.getBytes(StandardCharsets.UTF_8));
        }

        byte[] compressedVocab = TokenizerCoder.compressVocab(originalVocab);
        System.out.println("  Original vocab items: " + originalVocab.size());
        System.out.println("  Compressed vocab size: " + compressedVocab.length + " bytes");

        List<byte[]> restoredVocab = TokenizerCoder.decompressVocab(compressedVocab, originalVocab.size());
        System.out.println("  Restored vocab items: " + restoredVocab.size());

        if (originalVocab.size() != restoredVocab.size()) {
            throw new RuntimeException("Mismatch in vocab count!");
        }
        for (int i = 0; i < originalVocab.size(); i++) {
            if (!Arrays.equals(originalVocab.get(i), restoredVocab.get(i))) {
                throw new RuntimeException("Mismatch in vocab content at index " + i + "!");
            }
        }
        System.out.println("  [+] Vocab round-trip: SUCCESS (100% Match)");

        // 2. Test BPE Merges index pack/unpack
        System.out.println("\n[Test 2] BPE Merges Binary Index Coder...");
        List<int[]> originalMerges = new ArrayList<>();
        originalMerges.add(new int[]{1015, 2030});
        originalMerges.add(new int[]{45, 12});
        originalMerges.add(new int[]{16777215, 50000});
        originalMerges.add(new int[]{0, 1});
        originalMerges.add(new int[]{100000, 200000});

        byte[] compressedMerges = TokenizerCoder.compressMerges(originalMerges);
        System.out.println("  Original merges items: " + originalMerges.size());
        System.out.println("  Compressed merges size: " + compressedMerges.length + " bytes");

        List<int[]> restoredMerges = TokenizerCoder.decompressMerges(compressedMerges);
        System.out.println("  Restored merges items: " + restoredMerges.size());

        if (originalMerges.size() != restoredMerges.size()) {
            throw new RuntimeException("Mismatch in merges count!");
        }
        for (int i = 0; i < originalMerges.size(); i++) {
            if (originalMerges.get(i)[0] != restoredMerges.get(i)[0] ||
                originalMerges.get(i)[1] != restoredMerges.get(i)[1]) {
                throw new RuntimeException("Mismatch in merges content at index " + i + "!");
            }
        }
        System.out.println("  [+] Merges round-trip: SUCCESS (100% Match)");

        // 3. Test XOR-FEC Parity
        System.out.println("\n[Test 3] XOR-FEC Parity Calculation...");
        byte[] c1 = {(byte)0xAA, (byte)0xBB, (byte)0xCC, (byte)0xDD};
        byte[] c2 = {(byte)0x11, (byte)0x22, (byte)0x33, (byte)0x44};
        byte[] c3 = {(byte)0x55, (byte)0x66, (byte)0x77, (byte)0x88};
        List<byte[]> chunks = new ArrayList<>();
        chunks.add(c1);
        chunks.add(c2);
        chunks.add(c3);

        byte[] parity = TokenizerCoder.computeXorFecParity(chunks, 4);
        byte[] expectedParity = {
            (byte)(0xAA ^ 0x11 ^ 0x55),
            (byte)(0xBB ^ 0x22 ^ 0x66),
            (byte)(0xCC ^ 0x33 ^ 0x77),
            (byte)(0xDD ^ 0x44 ^ 0x88)
        };

        if (!Arrays.equals(parity, expectedParity)) {
            throw new RuntimeException("Mismatch in XOR-FEC parity!");
        }
        System.out.println("  [+] XOR-FEC computation: SUCCESS");

        System.out.println("\n=========================================================");
        System.out.println("  ALL JAVA TESTS PASSED SUCCESSFULLY!");
        System.out.println("=========================================================");
    }
}

File: foundations/UFO-Compression-Rust/test_tokenizer_coder.rs
mod tokenizer_coder;

fn main() {
    println!("=========================================================");
    println!("  RUNNING RUST UFO TOKENIZER CODER VERIFICATION");
    println!("  Watermark: ip zymatica.space");
    println!("=========================================================");

    // 1. Test Vocab Coder
    println!("\n[Test 1] Prefix-Suffix Vocab Coder...");
    let original_vocab: Vec<&[u8]> = vec![
        b"hello",
        b"hell",
        b"heaven",
        b"heavy",
        b"world",
        b"word",
        b"work",
        b"worker",
        b"working"
    ];

    let compressed_vocab = tokenizer_coder::compress_vocab(&original_vocab);
    println!("  Original vocab items: {}", original_vocab.len());
    println!("  Compressed vocab size: {} bytes", compressed_vocab.len());

    let restored_vocab = tokenizer_coder::decompress_vocab(&compressed_vocab, original_vocab.len());
    println!("  Restored vocab items: {}", restored_vocab.len());

    assert_eq!(original_vocab.len(), restored_vocab.len());
    for i in 0..original_vocab.len() {
        assert_eq!(original_vocab[i], restored_vocab[i]);
    }
    println!("  [+] Vocab round-trip: SUCCESS (100% Match)");

    // 2. Test BPE Merges Coder
    println!("\n[Test 2] BPE Merges Binary Index Coder...");
    let original_merges: Vec<(u32, u32)> = vec![
        (1015, 2030),
        (45, 12),
        (16777215, 50000),
        (0, 1),
        (100000, 200000)
    ];

    let compressed_merges = tokenizer_coder::compress_merges(&original_merges);
    println!("  Original merges items: {}", original_merges.len());
    println!("  Compressed merges size: {} bytes", compressed_merges.len());

    let restored_merges = tokenizer_coder::decompress_merges(&compressed_merges);
    println!("  Restored merges items: {}", restored_merges.len());

    assert_eq!(original_merges.len(), restored_merges.len());
    for i in 0..original_merges.len() {
        assert_eq!(original_merges[i], restored_merges[i]);
    }
    println!("  [+] Merges round-trip: SUCCESS (100% Match)");

    // 3. Test XOR-FEC Parity
    println!("\n[Test 3] XOR-FEC Parity Calculation...");
    let c1 = vec![0xAA, 0xBB, 0xCC, 0xDD];
    let c2 = vec![0x11, 0x22, 0x33, 0x44];
    let c3 = vec![0x55, 0x66, 0x77, 0x88];
    let chunks = vec![c1, c2, c3];

    let parity = tokenizer_coder::compute_xor_fec_parity(&chunks, 4);
    let expected_parity = vec![
        0xAA ^ 0x11 ^ 0x55,
        0xBB ^ 0x22 ^ 0x66,
        0xCC ^ 0x33 ^ 0x77,
        0xDD ^ 0x44 ^ 0x88
    ];
    assert_eq!(parity, expected_parity);
    println!("  [+] XOR-FEC computation: SUCCESS");

    println!("\n=========================================================");
    println!("  ALL RUST TESTS PASSED SUCCESSFULLY!");
    println!("=========================================================");
}
